syncapi/streams/stream_pdu.go

```diff
@@ -0,0 +1,303 @@
+package streams
+
+import (
+	"context"
+
+	"github.com/matrix-org/dendrite/syncapi/types"
+	userapi "github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+type PDUStreamProvider struct {
+	StreamProvider
+}
+
+func (p *PDUStreamProvider) Setup() {
+	p.StreamProvider.Setup()
+
+	p.latestMutex.Lock()
+	defer p.latestMutex.Unlock()
+
+	id, err := p.DB.MaxStreamTokenForPDUs(context.Background())
+	if err != nil {
+		panic(err)
+	}
+	p.latest = id
+}
+
+func (p *PDUStreamProvider) CompleteSync(
+	ctx context.Context,
+	req *types.SyncRequest,
+) types.StreamPosition {
+	from := types.StreamPosition(0)
+	to := p.LatestPosition(ctx)
+
+	// Get the current sync position which we will base the sync response on.
+	r := types.Range{
+		From:      to,
+		To:        0,
```

Comment:
This feels like it's the wrong way around?

Reply:
I've added a comment to this but it's right - for a complete sync, we want to show the most recent messages in the room, so we start at the current position and work backwards.